### ISSUE TITLE
At low and mid-zooms, merge "same" road segments

### ISSUE THREAD

**Issue by nvkelso** [mapzen/vector-datasource#358]
We're sending over a lot of chunky data in tiles at low- and mid- zooms which results in larger file sizes, longer time to render, and makes drawing pretty labels difficult (lots of colliding labels on small segments instead of spaced labels over longer runs.

For example, zoom 8:

![screen shot 2015-10-29 at 5 43 05 pm](https://cloud.githubusercontent.com/assets/853051/10833035/485f245a-7e4c-11e5-85ae-d241775188f4.png)

```
roads:
    data: { source: osm }
    filter: { not: { kind: [rail, ferry] } }
    draw:
        lines:
            order: 2
            # color: '#777'
            color: Style.color.randomColor()
            width: 2px
```
Download full [scene.txt](https://github.com/mapzen/vector-datasource/files/23796/scene.txt) file (rename txt to yaml) »

/cc @bcamper, @sensescape, @zerebubuth, discuss!

**Comment by zerebubuth** [comment]
Part of this is probably made worse by the landuse cutting, and #314 should address that. The other part of this is likely to be changes in status, e.g: network membership, name, tunnel, bridge, layer, and any of the other tags we're collecting (operator, type, colour, ascent, descent, etc...). We'll only be able to merge these if we're able to discard / ignore some of these tags. So...

@nvkelso Q: At which zoom do we no longer draw / care about changes in `tunnel`, `bridge`, layer, etc... status?

**Comment by zerebubuth** [comment]
For example, here's a roughly similar area with the roads coloured by a hash of their name, ref, kind and landuse_kind. There is a lot of improvement, because we're (effectively) dropping almost all the rest of the attributes. This picture represents an "upper bound" of what we can realistically do without putting a lot of effort into generalisation techniques (which would be awesome cool, though).

![nyc_z8](https://cloud.githubusercontent.com/assets/271360/10854164/ebed94fe-7f31-11e5-949e-c2b69a1b5d1c.jpg)

Here's the [scene.txt](https://github.com/mapzen/vector-datasource/files/24446/scene.txt). Note that the hash function is [from this SO answer](http://stackoverflow.com/questions/7616461/generate-a-hash-from-string-in-javascript-jquery#answer-7616484).

**Comment by bcamper** [comment]
Nice hash demo :)  My preference would be to start by choosing a set of attributes that allows us to get as close to this as possible (which does look like a significant improvement) while still preserving what we think is needed for styling. Then in the mid/longer-term do some more advanced generalization if possible (but if we're up for it, starting on that now would be fun too :)

**Comment by nvkelso** [comment]
Neat! We'll probably want slightly more chunky data in tiles (with slightly more attributes), and then Tangram will probably use a minimal attribute set to generate the label line geoms?

**Comment by zerebubuth** [comment]
If we're going to have linestring merging client-side anyway, then why not just jump straight to that?

Client-side would give us more flexibility about what to merge (e.g: drawing the stroke merging based on kind & landuse_kind, but labelling based on merging on name & ref), leading to better properties. This would have more immediate impact on styling and visual quality. The server-side implementation would then just be an optimisation to reduce file size and processing time on the client.

One question is whether we want to do merging across tile boundaries? It's futile to do this server-side since we clip to the tile boundary before encoding the tile, but it seems that it would improve label quality, especially near tile corners. However, I can see that making tile / geometry management much more hassle on the client.

**Comment by bcamper** [comment]
Well my motivation for wanting merging on the server-side is:

- Reduce the payload to client from all the duplicated properties.
- Improve responsiveness by reduce the amount of feature processing we need to do on the client during tile building (rather than add to it ;) It's already slow having to process so many lines, and while adding an up-front filter step may get back some of that time with an initial cull, obviously I'd prefer to not have to spend that time either.
- Simplify the features in a topology-aware way, especially culling all those 1-4 pixel geometries before we even waste client time processing them.
- Benefit other renderers besides Tangram.

I'm usually very pro-deferring to the client-side for greater flexibility, but I thought this was an area we could make significant improvements on the server in the near/mid-term, with the benefits of one-time cost to merge vs. on every client tile load.

**Comment by zerebubuth** [comment]
Sure. Sorry if it sounded like I was arguing against implementing it server-side, I should have been more explicit that it sounds to me like we'll need to do _both_. It looks like we've got a bunch of trade-offs:

1. The server must be conservative about merging, therefore more properties ⇒ worse merge, more lines, larger tiles, unhappy users.
2. However, culling properties server-side means fewer properties available for use in the style ⇒ less styling flexibility, unhappy cartographers.
3. Merging client-side adds complexity to the code, so more properties ⇒ slower tiles, battery drain, unhappy users.

We also have these things to take into account:

* We'll only get cross-tile-boundary merging on the client, but we might not even want this (perhaps it adds too much complexity).
* For the best results, we'll need to take the style into account (i.e: which properties are important), which is only known on the client.
* To avoid wasting time/battery on all clients, the server should merge as much as it safely can.

All of this leads me to believe that we are going to have to implement merging both client-side and server-side, and we'll have to try to minimise the number of properties at mid-to-low zoom levels.

There might be things that we can do on the server to make it easier to merge on the client, though? Perhaps pre-grouping similar roads (e.g: already sharing >80% of their properties) to cut down the search space on the client?

**Comment by bcamper** [comment]
That's a good summary, thanks :)  I agree we will probably end up doing some of both server and client-side. I could definitely be wrong, but my impression of these mid-zooms was that there were a lot of highway segments that were pretty similar property-wise, and we could hopefully get away with grouping on the properties that people care about for styling 80% of the time. Of course, if we investigate and find we can only get smallish savings in exchange for a lot of extra processing time, no need to go through with it. I was mostly interested in the initial research (like your screenshots above) to find out what our potential savings might be. I recall talking with @kevinkreiser about this too.

Regarding client-side merging across tiles.... yeah, gonna be a challenge :)  I've thought about it, but so much of the pipeline relies on indexing geometries and GL buffers to tiles (and ultimately you need *some* indexing/scene graph scheme for data management, and tiles are the natural one the data already comes in), it could be a big re-architecture relative to the gain. I think we may have more luck with styling options that help minimize breaks at tiles, such as hints in the data indicating when a line is cut at a tile (and on which end), so we could have smarter joins.

**Comment by zerebubuth** [comment]
Some preliminary results on this:

## Tile Size

This shows the reduction in the number of bytes in the whole tile for a TopoJSON format tile. The merging is only running on the roads layer, so there's some fixed overhead, but the reduction in mid-to-low zooms is pretty impressive.

```
 Z   Before ->    After (percent)
 6 13349803 -> 13349769 ( -0.0%)
 7  5745322 ->  5745328 (  0.0%)
 8 16326470 ->  3903376 (-76.1%)
 9 14035601 ->  6001753 (-57.2%)
10  9717532 ->  4090008 (-57.9%)
11  6163057 ->  2799254 (-54.6%)
12  4623569 ->  2379811 (-48.5%)
13  3123791 ->  2161468 (-30.8%)
14  3806721 ->  3153995 (-17.1%)
15  1071345 ->   856680 (-20.0%)
16   384357 ->   327120 (-14.9%)
17   180227 ->   180227 (  0.0%)
18    76519 ->    76519 (  0.0%)
```

## Item counts

This is where it gets more exciting! These counts are _only_ for the roads layer, and count the number of features (many are multi-polygons, and have separate parts - I can produce numbers on individual parts instead if that's of more interest). Between zooms 8 and 11 there's an order of magnitude reduction in the count, between 12 and 14 it's about a factor of 5.

```
 Z   Before ->    After (percent)
 6     1504 ->     1504 (  0.0%)
 7     1530 ->     1530 (  0.0%)
 8    56393 ->     1518 (-97.3%)
 9    36651 ->     1398 (-96.2%)
10    28487 ->      938 (-96.7%)
11    17908 ->      885 (-95.1%)
12    15003 ->     2254 (-85.0%)
13     6983 ->     1444 (-79.3%)
14     5184 ->      819 (-84.2%)
15     1998 ->      541 (-72.9%)
16      545 ->      158 (-71.0%)
17      167 ->      167 (  0.0%)
18       65 ->       65 (  0.0%)
```

There are some limitations currently around the configurability of the merge operation and the property keys it's using, and I'll fix those next.

Here's what the equivalent screenshot to the one which began this thread looks like now:

![nyc_z8_after_trim_3](https://cloud.githubusercontent.com/assets/271360/12127021/af7496a4-b3ec-11e5-8ae6-2cdb30cf172d.jpg)

**Comment by bcamper** [comment]
Amazing!! What properties are you merging on right now?

**Comment by nvkelso** [comment]
Huge progress, looking great :) What is the impact on tile generation time?

**Comment by zerebubuth** [comment]
> What properties are you merging on right now?

I'm dropping `is_bridge`, `is_tunnel`, `oneway`, `name`, `ref` and `network` at a bunch of different zooms and depending on the road `kind`. The full details are [in the code](https://github.com/mapzen/TileStache/blob/358-merge-road-segments/TileStache/Goodies/VecTiles/transform.py#L379-L401) and that's the bit I need to make more configurable.

> What is the impact on tile generation time?

I've not benchmarked yet, but anecdotally it's pretty bad. Bad enough that I noticed; it "felt" slower. Once I've got the code into shape then I'll run some benchmarking and see what we can do to mitigate / optimise.

**Comment by zerebubuth** [comment]
Contrary to my previous statement, in fact the tile generation got _faster_. Here's the full set of data:

## Feature Counts

```
 Z   Before ->    After (percent)
 6      214 ->      214 (  0.0%)
 7      218 ->      218 (  0.0%)
 8     8056 ->      216 (-97.3%)
 9     5235 ->      199 (-96.2%)
10     4069 ->      134 (-96.7%)
11     2558 ->       97 (-96.2%)
12     2143 ->      297 (-86.1%)
13      997 ->      189 (-81.0%)
14      740 ->      115 (-84.5%)
15      285 ->       77 (-72.9%)
16       77 ->       22 (-71.0%)
17       23 ->       23 (  0.0%)
18        9 ->        9 (  0.0%)
```

## Uncompressed Tile Sizes
```
 Z   Before ->    After (percent)
 6  1907133 ->  1907106 ( -0.0%)
 7   820762 ->   820762 ( -0.0%)
 8  2332353 ->   557706 (-76.1%)
 9  2005132 ->   857379 (-57.2%)
10  1388260 ->   584272 (-57.9%)
11   880505 ->   391521 (-55.5%)
12   660666 ->   334758 (-49.3%)
13   446421 ->   305935 (-31.5%)
14   544032 ->   450352 (-17.2%)
15   153456 ->   122791 (-20.0%)
16    55012 ->    46835 (-14.9%)
17    25747 ->    25747 (  0.0%)
18    10932 ->    10932 (  0.0%)
```

## Time (in seconds) to Download Tile

```
 Z   Before ->    After (percent)
 6     15.6 ->     15.6 (  0.2%)
 7      7.8 ->      7.8 ( -0.3%)
 8     22.4 ->     16.7 (-25.5%)
 9     25.1 ->     21.5 (-14.4%)
10     12.7 ->     10.0 (-20.9%)
11      7.6 ->      6.0 (-21.2%)
12      5.6 ->      4.9 (-12.4%)
13      3.7 ->      3.4 ( -9.1%)
14      4.2 ->      4.1 ( -3.6%)
15      1.3 ->      1.3 ( -4.4%)
16      0.6 ->      0.6 ( -0.9%)
17      0.4 ->      0.4 (  2.8%)
18      0.3 ->      0.3 ( -4.7%)
```

It should be noted that there's plenty of noise in the last result, particularly at higher zooms where the time is short and more easily swamped with other (e.g: network) effects.

**Comment by nvkelso** [comment]
Minor refinements suggested in the #468 but otherwise good :)

**Comment by nvkelso** [comment]
Going to hold off looking at this one until we verify the rest of v0.7-dev is working.

**Comment by nvkelso** [comment]
@bcamper Can you take a look at dev and give :+1: or :-1: ?

```
sources: 
    url: '//vector.dev.mapzen.com/osm/all/{z}/{x}/{y}.topojson'
```

Look especially at complex multiline strings (was mostly single line strings before).

**Comment by migurski** [comment]
I made [a weird service that merged “same” roads and routes](https://openstreetmap.us/~migurski/streets-and-routes/), and I think it’s pretty important that we do this kind of data pre-processing for low zoom levels. Below z12 or so, OSM isn’t really OSM anymore, since the editing workflow and project conceptually designed around things observable from z14-z18. New derived datasets to bridge the gap between lowest OSM and highest Natural Earth would be cool.

**Comment by bcamper** [comment]
Totally, it's that mid-range where things get weird.

**Comment by nvkelso** [comment]
OMG, "dual carriageways"!

**Comment by migurski** [comment]
![dual](https://cloud.githubusercontent.com/assets/58730/12311290/1ff5c14a-ba0b-11e5-9fa1-9d289adb95d4.jpg)

**Comment by zerebubuth** [comment]
Agree, generalising is a Hard Problem, but one that's necessary for both operational and display reasons. Operational because smaller tiles, better cache hit ratio, better framerate. Display because `ETOOMUCHDETAIL`.

That being said, I think it's important that we don't throw away the advantages of OSM just because generalisation is Hard. We're living in the future and I totally believe that we can make generalisation work with minutely updates (see hand-waving about NEAT). Rather than bridge the gap between OSM and NE with a 3rd dataset, I'd like to see our tech become so good that we can use OSM for all zoom levels.

**Comment by bcamper** [comment]
I haven't dug into the data yet, but seeing this Eraser Map difference. It looks like the features are still there, but the styling is not matching in the same way. This could either be a data issue, or a styling one.

Production:
![screen shot 2016-01-14 at 3 40 24 pm](https://cloud.githubusercontent.com/assets/16733/12336935/32c6657a-bad5-11e5-9b68-a9b74588ad8c.png)

Dev:
![screen shot 2016-01-14 at 3 40 32 pm](https://cloud.githubusercontent.com/assets/16733/12336936/32c79b0c-bad5-11e5-8545-ec12a85d0346.png)

**Comment by bcamper** [comment]
Note the above scene resulted in a **33% decrease** in geometry build time (due to decreased filter checks) and **8% decrease** (-30k polygons) in the amount of geometry generated for the tile (due to decreased line caps). I was hoping for even better but that is still a big improvement!

**Comment by nvkelso** [comment]
Looks like the change in style is caused by dropping `ref` from major_road features too early:

https://github.com/mapzen/vector-datasource/pull/468#discussion_r49804593

**Comment by nvkelso** [comment]
**tl;dr**
We should stop including `motorway_link` at zooms 8, 9, and 10. That'll reduce the feature count quite a bit (probably between 10% and 30% more)! That's a small enough change let's include it with this scope of work, see below for proposed fix.

Besides the missing `ref` bit covered in https://github.com/mapzen/vector-datasource/pull/468#discussion_r49804593 that we need to fix, this is looking good.

===

I loaded up the [scene.txt](https://github.com/mapzen/vector-datasource/files/23796/scene.txt) file (rename txt to yaml) in Tangram Play to get these before / after images:

**new (dev):**
![screen shot 2016-01-14 at 16 09 42](https://cloud.githubusercontent.com/assets/853051/12342197/119b565a-badd-11e5-9e7b-239931bcca9c.png)

**old (prod):**
![screen shot 2016-01-14 at 16 09 36](https://cloud.githubusercontent.com/assets/853051/12342198/119d63dc-badd-11e5-9c0f-4734edf844be.png)

It looks like the remaining sprinkling of tiny road segments are `motorway_link` features in the `kind: highway` grouping:

http://vector.dev.mapzen.com/osm/all/10/300/385.topojson

In issue #122 we [promoted](https://github.com/mapzen/vector-datasource/pull/123/files#diff-85235d701f80586d926a4211261dbbc1R96) those `motorway_link` features up to zoom 9 (they had been zoom 12). In retrospect, that was probably too early. Showing `motorway_link` features at zoom 11 still makes a lot of sense visually. Zoom 10 seems marginal visually, but the savings of dropping them from the input at zoom 9 and zoom 10 seems to have huge savings in feature count and overall perf. (A later commit must have moved `motorway_link` features up to zoom 8, which is certainly too early!)

https://github.com/mapzen/vector-datasource/blob/master/data/functions.sql#L275-L281

```
CREATE OR REPLACE FUNCTION mz_calculate_highway_level(highway_val text, service_val text)
RETURNS SMALLINT AS $$
BEGIN
  RETURN CASE
    WHEN highway_val IN ('motorway', 'trunk', 'motorway_link', 'primary')     THEN  8
    WHEN highway_val IN ('secondary')                                         THEN 10
    WHEN highway_val IN ('tertiary')                                          THEN 11
```

Adjusting like so should be good:

```
    WHEN highway_val IN ('motorway', 'trunk', 'primary')               THEN  8
    WHEN highway_val IN ('secondary')                                  THEN 10
    WHEN highway_val IN ('tertiary', 'motorway_link')                  THEN 11
```

**Background reading:** `motorway_link` features are part of the general `kind: highway` set here:

https://github.com/mapzen/TileStache/blob/integration-1/TileStache/Goodies/VecTiles/transform.py#L124-L127

```
road_kind_highway = set(('motorway', 'motorway_link'))
road_kind_major_road = set(('trunk', 'trunk_link', 'primary', 'primary_link',
                            'secondary', 'secondary_link',
                            'tertiary', 'tertiary_link'))
```

Here's similar view to Brett's screenshots:

<img width="1341" alt="screen shot 2016-01-14 at 16 51 23" src="https://cloud.githubusercontent.com/assets/853051/12342462/5f761138-badf-11e5-863b-36cbea14f83a.png">

Reviewing one of those tiles:

* http://vector.dev.mapzen.com/osm/all/9/150/192.topojson

There are 70 `motorway_links` out of 320 road features (`motorway`,`motorway_link`,`trunk`, etc) features or 21% of the total features, and around 25% of `motorway*` features).

Back in the bay area:

* http://vector.dev.mapzen.com/osm/all/8/40/98.topojson

There are 13 `motorway_links` out of 89 road features (`motorway`,`motorway_link`,`trunk`, etc) features or 15% of the total features, and around 33% of `motorway*` features).

**Comment by zerebubuth** [comment]
I've added #488 to track the `motorway_link` issue and #489 for the `ref` issue.

I find it quickly gets pretty confusing when issues are mentioned in comments (other than in open PRs) and it's easy for me to lose track of these unless there's a separate "handle" to track them. So it would help me a lot if these were added as new issues rather than widening the scope of existing issues.

**Comment by nvkelso** [comment]
LGTM

**Comment by nvkelso** [comment]
For posterity, the scene file above no longer works, but this snippet does:

```
   # useful to example road geometry merging & Tangram labeling w/r/t tile boundaries
   z-randomColor:
       data: { source: mapzen, layer: roads }
       draw:
           lines:
               order: 1000
               # color: '#777'
               color: |  #Style.color.randomColor()
                   function() {
                       var letters = '0123456789ABCDEF';
                       var color = '#';
                       for (var i = 0; i < 6; i++ ) {
                           color += letters[Math.floor(Math.random() * 16)];
                       }
                       return color;
                   }
               width: [[13,0.5px],[14,1px],[16,2px]]
```
